src/parser.rs

```diff
@@ -216,11 +263,56 @@ impl<'i> Iterator for Input<'i> {
     }
 }
 
+/// Wrapper for syntax violation callback functions.
+#[derive(Copy, Clone)]
+pub enum ViolationFn<'a> {
+    NewFn(&'a (Fn(SyntaxViolation) + 'a)),
+    OldFn(&'a (Fn(&'static str) + 'a)),
+    NoOp
+}
+
+impl<'a> ViolationFn<'a> {
+    pub fn call(self, v: SyntaxViolation) {
+        match self {
+            ViolationFn::NewFn(f) => f(v),
+            ViolationFn::OldFn(f) => f(v.description()),
+            ViolationFn::NoOp => {}
+        }
+    }
+
+    pub fn call_if<F>(self, v: SyntaxViolation, test: F)
```

Comment:
Please just use `if` where you currently call `call_if`.

Reply:
`Violation#call_if` is direct replacement for the original `Parser#syntax_violation_if`.  I _think_ this was introduced as and remains a useful optimization and convenience in that it does an early check for `NoOp` (short-circuit eval of test). Beyond that I have been trying not to change every call by needing to rewrite with `if` blocks, which would be a much more extensive change of the `Parser`, and require more scrutiny.

While I could have kept and just modified the original Parser method in place, I think this PR better collects those concerns under `ViolationFn`.  I added comments to these `ViolationFn` methods to hopefully clear this up. Hope that is sufficient.